### PR TITLE
#1070 #1071 Search plugin(2)

### LINKED ISSUE #1071: Programs: Treatment Supporter
From: https://github.com/openmsupply/open-msupply/issues/644#issue-1383248595

> #### `Treatment Supporter`:
> 
>   - should be able to specify if TS is same as `Next of Kin` from the registration details (avoid unnecessary data entry). That would require to pull NoK data from the registration form into the enrolment if that checkbox is ticked.
>   - TS can be different from NoK cause sometimes patient do not want to share health data with their family

Related to #1070 in that we'll probably want to create a general component that can handle both these scenarios.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #1071 (and #1070)

# 👩🏻‍💻 What does this PR do? 
 

Continuation of PR #1223 -- adds the other "Search by Document" component, which looks up a path specified in an existing document (currently only "currentPatient" or "currentEncounter")

This component has no user interaction, it just displays the referenced value (according to a "displayString" option).

The options are similar to the other component, so the overall "Search" widget should have similar config in either case.

Have also made a few improvements to the "SearchWithUserSource" component.

# 🧪 How has/should this change been tested? 

Implemented in "HIV Care and Treatment" enrolment form for "Treatment supporter". There is a checkbox where the user can specify if the "Treatment supporter" is the same person as the next of kin. If "true", this component is used to fetch that data from the current patient and display it. If not, the user must fill out standard text input fields to enter the data (but the same data node will be updated in either case)

Unchecked:
Screenshot 2023-02-23 at 12 35 27 PM

Checked:
Screenshot 2023-02-23 at 12 36 34 PM

## 💌 Any notes for the reviewer?

Have tested error states as much as I can, but try messing around with the options specified in the UI schema to see the result (including errors)

See inline comments for additional details.

### CODE REVIEW

File: client/packages/common/src/utils/formatters/JSXFormatters.tsx [hunk 1]
```diff
@@ -6,12 +6,14 @@ export const JSXFormatters = {
...
-    ));
+    return lines.map((line, i) =>
+      line !== '' ? (
+        <span key={i}>
+          {line}
+          {/* Don't add line break to final line */}
+          {i === lines.length - 1 ? null : <br />}
+        </span>
+      ) : null
+    );
```
Comment (CarlosNZ) [inline client/packages/common/src/utils/formatters/JSXFormatters.tsx:17]: Small change so that empty lines don't take up space (they won't get a line break)

File: client/packages/programs/src/JsonForms/common/JsonForm.tsx [hunk 2]
```diff
@@ -152,6 +155,7 @@ const FormComponent = ({
           setError?.(false);
         }
       }}
+      ajv={handleDefaultsAjv}
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/JsonForm.tsx:158]: @clemens-msupply -- didn't end up using this as we discussed, but do you think it's okay to leave the functionality in place so "default" values can be used when we want them?

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    Options,
+    uischema.options
+  );
+  const { currentPatient = {} } = usePatientStore();
+
+  const {
+    data: currentEncounter,
+    isLoading,
+    isError,
+  } = useEncounter.document.byId(useEncounter.utils.idFromUrl());
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:66]: Current encounter only available when we're in an Encounter form, so that the encounter ID can be fetched from URL. Will report an error (see below) if the specified document (e.g. "encounter") is unavailable in the current context.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  }, [requestedData]);
+
+  const displayElement = (
+    <Typography>
+      {options?.displayString ? (
+        replaceHTMLlineBreaks(
+          formatTemplateString(options.displayString, requestedData ?? {}, '')
+        )
+      ) : (
+        <pre>{JSON.stringify(requestedData, null, 2)}</pre>
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:101]: There's no reasonable default for this (since data could be pulled from any node of the Patient or Encounter object), so will just display raw object as a string if no display string specified.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    ? zErrors
+    : formErrors
+    ? formErrors
+    : isLoading
+    ? null
+    : isError || !documentData
+    ? t('control.search.error.no-document')
+    : !requestedData
+    ? t('control.search.error.no-data', { docPath: options?.path })
+    : null;
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:116]: Bit ugly, but hopefully makes sense.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx [hunk 6]
```diff
@@ -20,33 +21,36 @@ import { useSearchQueries, QueryValues } from './useSearchQueries';
...
+    /**
+     * List of fields to save in document data (from selected item object)
+     */
+    saveFields: z.array(z.string()).optional(),
+    /**
+     * Text to show in input field before user entry
+     */
+    placeholderText: z.string().optional(),
+  })
+  .strict();
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:48]: Just made it `strict` and included the required "source" value.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx [hunk 7]
```diff
@@ -80,16 +84,19 @@ export const SearchWithUserSource = (props: ControlProps) => {
...
-      field => (newObj[field] = selectedResult[field] ?? null)
-    );
-    handleChange(path, newObj);
+    else {
+      const newObj: Record<string, any> = {};
+      saveFields?.forEach(
+        field => (newObj[field] = selectedResult[field] ?? null)
+      );
+      handleChange(path, newObj);
+    }
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:93]: Forgot to return early from the "no saveFields condition", which caused problems, so changed to an if/else.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx [hunk 8]
```diff
@@ -80,16 +84,19 @@ export const SearchWithUserSource = (props: ControlProps) => {
...
     setEditMode(false);
   };
 
-  if (!options) return null;
+  const error = zErrors ?? props.errors ?? queryError ?? null;
+
+  if (!visible) return null;
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:99]: I noticed quite a few of the existing components aren't handling `visible` (or `enabled`), so have made an issue to check and fix these: #1228

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx [hunk 9]
```diff
@@ -91,17 +96,9 @@ export const useSearchQueries = ({
...
-        'code',
-        'firstName',
-        'lastName',
-        'dateOfBirth',
-        'gender',
-        'email',
-        'document',
-      ],
-      placeholderText: placeholderText ?? 'Search by patient code...',
+      saveFields: saveFields ?? null,
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx:99]: Seemed like "save the whole object" would be a better default if `saveFields` not specifically defined.

**Comment by clemens-msupply** [comment]
When adding the person there is actually too much information added, e.g. we shouldn't add the contact details or other document data:

image
image

There is also a `code: null` entry... Could you please clean up the data that is added to the document?

**Comment by clemens-msupply** [comment]
Suggestion: when the treatment supporter checkbox is is ticked the display view should look the same as for manual entry, i.e. have multiple inputs for:
- firstname, lastname, phone, mobile, email
- but all of these disabled for user input

In this way the UI doesn't "jump" when ticking the box and it is clearer what happens.

**Comment by clemens-msupply** [comment]
When there is no next of kin for a patient this happens:

image

Maybe disable/hide the tickbox or show message if there is no next of kin?

File: client/packages/programs/src/JsonForms/common/JsonForm.tsx (hunk 2 above)
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/common/JsonForm.tsx:158]: Should we remove till we need it? might be confusing otherwise

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+import { useZodOptionsValidation } from '../../common/hooks/useZodOptionsValidation';
+import { usePatientStore } from 'packages/programs/src/hooks';
+import { useEncounter } from '../../../api';
+import { RegexUtils, JSXFormatters } from '@openmsupply-client/common';
+
+const Options = z
+  .object({
+    source: z.enum(['document']),
+    /**
+     * Which pre-defined query to use (in useSearchQueries)
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:24]:
```suggestion
     * Source document type. Either the current patient or the current encounter (if applicable).
```

May `current` instead of `encounter` because you might not be in an encounter. Than it would also work for program enrollments

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    /**
+     * Path in the specified document to extract
+     */
+    path: z.string(),
+    /**
+     * Pattern for formatting selected result (as above)
+     */
+    displayString: z.string().optional(),
+    /**
+     * List of fields to save in document data (from selected item object)
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:36]:
```suggestion
     * List of fields to be saved in the document data (from selected source document)
     * The fields are stored in an object at the `scope` path of this UI control.
```

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx (hunk 3 above)
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:66]: as in comment above you could just use the "current" document

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx [hunk 12]
```diff
@@ -20,33 +21,36 @@ import { useSearchQueries, QueryValues } from './useSearchQueries';
...
-   */
-  saveFields: z.array(z.string()).optional(),
-  /**
-   * Text to show in input field before user entry
-   */
-  placeholderText: z.string().optional(),
-});
+const Options = z
+  .object({
+    source: z.enum(['user']),
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:26]: Think `name` or `patient` would be better?

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx [hunk 13]
```diff
@@ -20,33 +21,36 @@ import { useSearchQueries, QueryValues } from './useSearchQueries';
...
+    /**
+     * Pattern for formatting options list items (e.g. "${firstName} ${lastName}")
+     */
+    optionString: z.string().optional(),
+    /**
+     * Pattern for formatting selected result (as above)
+     */
+    displayString: z.string().optional(),
+    /**
+     * List of fields to save in document data (from selected item object)
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:40]:
```suggestion
     * List of fields to be saved in document data (from found patient)
```

**Review by clemens-msupply — CHANGES_REQUESTED** [review]

File: client/packages/programs/src/JsonForms/common/JsonForm.tsx (hunk 2 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/common/JsonForm.tsx:158]: I would expect default values to be fairly standard functionality, even if we're not currently utilising it. Have added an explanatory comment: 72247bef8

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx (hunk 3 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:66]: Yeah, that would be good, but wasn't sure how to access the generic "current" document. Any suggestions?

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx (hunk 12 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:26]: "user" means the search parameter comes from user input, as opposed to a document source. It's not talking about _what_ is being searched, but _where_ the "variable" for the search comes from. Hence 'user' and 'document' - but we could use that input to search a document, a patient, or anything.

**Comment by CarlosNZ** [comment]
> When adding the person there is actually too much information added, e.g. we shouldn't add the contact details or other document data:

Sure, it's currently just saving the whole "Person" object as I haven't specified any "saveFields", as I wasn't sure which ones we'd want. I guess I'll just add a few of the basic ones for now and wait till we have more specific requirements.

**Comment by CarlosNZ** [comment]
> When adding the person there is actually too much information added, e.g. we shouldn't add the contact details or other document data:

Where are you seeing this? The data auto-saved from the "Next of kin" information only has the captured "next of kin" fields (from the Patient document) -- i.e. firstName, lastName, relationship, phone, mobile, email.

**Comment by CarlosNZ** [comment]
> Suggestion: when the treatment supporter checkbox is is ticked the display view should look the same as for manual entry, i.e. have multiple inputs for:
> 
>     * firstname, lastname, phone, mobile, email
> 
>     * but all of these disabled for user input
> 
> 
> In this way the UI doesn't "jump" when ticking the box and it is clearer what happens.

Sure, I can leave the original input fields visible but disabled. But there's not really a simple way to re-populate them with the extracted data, and I'm not sure leaving *empty* disabled fields is a good idea. 

Or did you mean something else?

**Comment by CarlosNZ** [comment]
> Maybe disable/hide the tickbox or show message if there is no next of kin?

We'd need to write a visibility rule that can extract info from the current patient *before* any user input, but we don't currently have a way to do that. (rule scopes only refer to the current document data). 

(This would actually be pretty simple with a FigTree expression, so something to think about down the line...) 

Or did you mean something slightly different?

**Comment by clemens-msupply** [comment]
> > When adding the person there is actually too much information added, e.g. we shouldn't add the contact details or other document data:
> 
> Where are you seeing this? The data auto-saved from the "Next of kin" information only has the captured "next of kin" fields (from the Patient document) -- i.e. firstName, lastName, relationship, phone, mobile, email.

When searching the mother, the added mother has the full document data added...

**Comment by clemens-msupply** [comment]
> > Suggestion: when the treatment supporter checkbox is is ticked the display view should look the same as for manual entry, i.e. have multiple inputs for:
> > ```
> > * firstname, lastname, phone, mobile, email
> > 
> > * but all of these disabled for user input
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In this way the UI doesn't "jump" when ticking the box and it is clearer what happens.
> 
> Sure, I can leave the original input fields visible but disabled. But there's not really a simple way to re-populate them with the extracted data, and I'm not sure leaving _empty_ disabled fields is a good idea.
> 
> Or did you mean something else?

If you set the treatment supporter fields in the document data, JSONForms should display it automatically. You only need a rule to disable the input fields when the isTreatmentSupporter box is ticked

**Comment by CarlosNZ** [comment]
> If you set the treatment supporter fields in the document data, JSONForms should display it automatically. You only need a rule to disable the input fields when the isTreatmentSupporter box is ticked

I can get the original input fields to be disabled no problem. But what I mean is that if these input fields are not then populated with the auto-extracted Supporter data (which is currently displayed as text), it looks wrong. And doing so is quite a non-trivial thing, I think.

**Comment by clemens-msupply** [comment]
> > If you set the treatment supporter fields in the document data, JSONForms should display it automatically. You only need a rule to disable the input fields when the isTreatmentSupporter box is ticked
> 
> I can get the original input fields to be disabled no problem. But what I mean is that if these input fields are not then populated with the auto-extracted Supporter data (which is currently displayed as text), it looks wrong. And doing so is quite a non-trivial thing, I think.

you mean if not treatment supporter is set? maybe unset existing fields then?

**Comment by CarlosNZ** [comment]
> you mean if not treatment supporter is set? maybe unset existing fields then?

Probably easier if I just show you what I mean. I might not be understanding your suggestion properly, either.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx (hunk 5 above)
Comment (mark-prins) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx:114]: this is really hard to read - can you reformat as a method / switch statement / something.
I would go with a switch, that makes each condition clear - it's very hard to quickly follow nested ternary operators like this.

**Comment by CarlosNZ** [comment]
> > > When adding the person there is actually too much information added, e.g. we shouldn't add the contact details or other document data:
> > 
> > 
> > Where are you seeing this? The data auto-saved from the "Next of kin" information only has the captured "next of kin" fields (from the Patient document) -- i.e. firstName, lastName, relationship, phone, mobile, email.
> 
> When searching the mother, the added mother has the full document data added...

Oh right, you're talking about the Mother search one. Sure, I've added a "saveFields" property to limit it to: `["id", "code", "firstName", "lastName", "dateOfBirth", "gender"]`. Sound okay?

962a3748e

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx (hunk 5 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx]: Sure, I made a small function with fall-through early returns: ba721005e (Switch didn't really work as there's not a clear single switch expression to compare against)

In general I don't find nested ternaries too hard to follow (they basically just fall through until something is true, then return the next thing), but I agree that this is a particularly gnarly one.

**Comment by CarlosNZ** [comment]
> When there is no next of kin for a patient this happens:
> 
> image

I get this for a patient created using the forms:
Screenshot 2023-03-01 at 10 55 19 AM

Which, admittedly is not great either, but not sure how to improve without checking each sub-field for `undefined`, but we don't know in advance what the fields to check would be.

I think hiding the checkbox based on some (as yet uncertain) condition is probably the best solution, but we don't yet have the ability to do that.

**Review by clemens-msupply — COMMENTED** [review]
Works as expected now, thanks!

I made another comment about `z.discriminatedUnion` after getting confused about it (even after you told me ) I think it might be clearer to use it?

For the empty treatment supporter problem. I think we can still just fill it into the existing but disabled Person fields. Lets discuss later...

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx (hunk 12 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:26]: > Nit: the 'document' type is not really a search component, is it? but mare a "clone" component?

Yeah, I'd call it a "lookup" component, strictly speaking. :)

File: client/packages/system/src/Encounter/DetailView/DetailView.tsx [hunk 14]
```diff
@@ -34,6 +34,8 @@ export const DetailView: FC = () => {
     isError,
   } = useEncounter.document.byIdPromise(id);
 
+  console.log('encounter', encounter);
+
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/system/src/Encounter/DetailView/DetailView.tsx]: This is already gone -- are you looking at an older commit?

**Comment by CarlosNZ** [comment]
> miss the Rust's if expression here :) other alternative is to put it into a helper method or inline arrow function that returns the result. This would avoid a mutable let but not sure if its really nicer...

I already changed this as per @mark-prins ' request. Are you using an older commit? ba721005e

File: client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx (hunk 5 above)
Comment (mark-prins) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx]: 🙏  I do find it easier to read

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx (hunk 12 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:26]: > Ah ok I see, you mentioned it. Could you please use a `z.discriminatedUnion` in the parent component though and call 'source' `type`? (there is an existing usage of it which uses type as well). I think it becomes clearer in the code then... You also only have to call zod once in the main component then.

@clemens-msupply I just tried making a zod discriminated union at the top level. It works fine at that level and simplifies the code in the children. But the problem is I still need a named type of each union element in each of the children, otherwise the child components don't know *which* of the union contributor types to expect, and that gets really messy. Also, passing the pre-validated "options" to each child is harder and (even assuming I have two different child types) you have to put an `as` as you pass each "options" down, and I find this really messy and unsatisfying.

That's why I thought just defining a very loose type at the top level and leaving each child to do strict validation on their specifics was simpler.

I can show you what I mean if you still want to pursue it.

Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx]: @clemens-msupply actually I can do it better if I make specific types for each child type and then use them in the discriminated union.

I think this does the trick: c707afb5d

File: client/packages/programs/src/JsonForms/components/Search/Search.tsx [hunk 15]
```diff
@@ -4,15 +4,54 @@ import { withJsonFormsControlProps } from '@jsonforms/react';
...
+import { SearchWithDocumentSource } from './SearchWithDocumentSource';
+import { QueryValues } from './useSearchQueries';
 
-const Options = z.object({
+const UserOptions = z.object({
   /**
    * Source of the search data -- user input or extract it from document
    */
-  source: z.enum(['user', 'document']),
+  source: z.literal('user'),
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/Search.tsx:14]:
```suggestion
  type: z.literal('user'),
```

Think `type` is a better discriminator. Its also use elsewhere

Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/Search.tsx:14]: I don't really agree. "Type" is vague and could mean anything (and doesn't makes sense with "user"), whereas "source" is what is actually is.

Plus in the ui_schema, we already have "type" to describe what type of element it is ("Control", "Search", etc.) It's confusing to have "type" as an option as well.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx (hunk 12 above)
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx]: cool awesome yes that looks good!!

File: client/packages/programs/src/JsonForms/components/Search/Search.tsx (hunk 15 above)
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/Search/Search.tsx:14]: I meant like "Search type", e.g. type = "Document search" and "User search"

but yeah source would work as well. Type is more common as a discriminator though I would say. 

Yes looks slightly odd :)
``` js
{
  type: "Search"
  options: {
     type: "Document"
  }
}
```

don't really mind

**Review by clemens-msupply — APPROVED** [review]
